trinity/utils/filesystem.py

```diff
@@ -20,3 +21,41 @@ def is_under_path(base_path: str, path: str, strict: bool=True) -> bool:
     absolute_base_path = os.path.abspath(base_path)
     absolute_path = os.path.abspath(path)
     return absolute_path.startswith(absolute_base_path)
+
+
+class PidFile:
+    """
+    This class tracks the process id of the currently running process in a file.
+    Use it as a context manager so that the file gets cleaned up in a graceful shutdown.
+    There are some corner cases that may cause the pidfile to be left behind with a dead process,
+    but any running process within the context of a PidFile should always have an
+    associated pid file.
+
+    It happens to usually prevent running two processes with the same process_name
+    at the same time, but does not guarantee this behavior.
+    """
+    def __init__(self, process_name: str, path: Path) -> None:
+        self.filepath = path / (process_name + '.pid')
+        self._running = False
+
+    def __enter__(self) -> None:
+        if self.filepath.exists():
+            raise FileExistsError(
+                "File %s already exists -- cannot run the same process twice" % self.filepath
```

Comment:
I'm not entirely sure if this is really what we want. Isn't it totally fine to run multiple instances of Trinity in parallel? Like, what would be wrong with running a Ropsten and a Mainnet node in parallel (assuming my machine has the power to handle it)?

Reply:
The `PidFile` is located in the `data_dir` so these would be under separate paths if you ran ropsten & mainnet (so I don't think this is an issue).

It would however prevent you from running `trinity` and `trinity --light`

Reply:
> It would however prevent you from running trinity and trinity --light

Which I think is a good thing. I am not confident that that's safe to do.

Reply:
>The `PidFile` is located in the `data_dir`

Ah, I missed that. Yes, that sounds sane :+1: